src/common/type/stringId.h

```diff
@@ -0,0 +1,123 @@
+{
+    stringIdBit5 = 0,                                               // 5-bit encoding for a-z, 1-4, and - characters
+    stringIdBit6 = 1,                                               // 6-bit encoding for a-z, 0-9, A-Z, and - characters
+} StringIdBit;
+
+/***********************************************************************************************************************************
+Macros to define constant StringIds. ALWAYS use strIdGenerate() to create these macros. The parameters in the macros are not
+verified against each other so the str parameter is included only for documentation purposes.
+***********************************************************************************************************************************/
+#define STRID5(str, strId)                                          strId
+#define STRID6(str, strId)                                          strId
+
+/***********************************************************************************************************************************
+Functions
+***********************************************************************************************************************************/
+// Convert N chars to StringId, If the string is longer than the allowable length for the selected encoding then the StringID will
+// be marked as "partial" and will have a '+' appended whenever it is converted back to a string. This is to distiguish it from a
+// string with the same number of encoded characters that did not overflow.
+StringId strIdFromZN(const StringIdBit bit, const char *const buffer, const size_t size);
+
+// Convert String to StringId using strIdFromZN()
+__attribute__((always_inline)) static inline StringId
+strIdFromStr(const StringIdBit bit, const String *const str)
+{
+    return strIdFromZN(bit, strZ(str), strSize(str));
+}
+
+// Convert zero-terminted string to StringId using strIdFromZN()
+__attribute__((always_inline)) static inline StringId
+strIdFromZ(const StringIdBit bit, const char *const str)
+{
+    return strIdFromZN(bit, str, strlen(str));
+}
+
+// Write StringId to characters without zero-terminating. The buffer at ptr must have enough space to write the entire StringId,
+// which could be eight characters. However, the caller may know the exact (or max length) in advance and act accordingly. The
+// actual number of bytes written is returned.
+size_t strIdToZN(StringId strId, char *const buffer);
+
+// Convert StringId to String
+String *strIdToStr(const StringId strId);
+
+// Convert StringId to zero-terminated string. See strIdToZN() for buffer sizing and return value.
+size_t strIdToZ(const StringId strId, char *const buffer);
+
+/***********************************************************************************************************************************
+Generate constant StringIds
+
+To generate a constant StringId call strIdGenerate() in any debug build. It will throw an error with the generated StringId macro
+in the error message.
```

Comment:
I'm a little confused about the strIdGenerate. 

`strIdGenerate("mytest0123abc")` produces `STRID5("mytest0123abc", 0x107dde02932d32d4)`
but then I run `strIdToStr(0x107dde02932d32d4)` and I get `mytest!123ab+`

Since `mytest0123abc` contains a 0 I was thinking it would use STRID6, how do I get it to not replace the `0` with a `!` ?

Reply:
It should have used STRID6(). There was problem with the boundary checking, so one extra character worked, but more than one extra did not work. Fixed in 196c2b047a267.